### PR TITLE
feat: 1840 validator accepts replacement character in stop name field

### LINKED ISSUE #1840: Validator Accepts Replacement Character in `stop_name` Field
### Describe the problem

Mobility Data Validator currently accepts the Unicode replacement character (`�`) in the `stop_name` field. This character typically indicates a decoding error and should not be present in valid data. While I've specifically noticed the acceptance of this replacement character, I am unsure if other similar problematic characters are also being accepted. Ideally, the validator should flag or reject `stop_name` fields containing the replacement character and potentially other similar invalid characters, as was done by the older GTFS validator, FeedValidator - https://github.com/google/transitfeed/tree/master, which raised a Unicode error (E33) for invalid values in the stop_name field. Implementing similar checks in future releases would help prevent data issues caused by improperly decoded or malformed input. Could this behavior be reviewed and corrected in future releases please? Thank you for your consideration and for your continued work on this project! 

### Describe the new validation rule

replacement_char = "\uFFFD"
if replacement_char in stop_name:
	trigger_error("invalid_character_in_stop_name", stop_name)

### Sample GTFS datasets

[auckland_gtfs.zip](https://github.com/user-attachments/files/16991535/auckland_gtfs.zip)

### Severity

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by qcdyx**
**Summary:**
Closes #1840 

**Expected behavior:** 
If stop name contains a replacement character, show a notice.
image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by qcdyx** [comment]
@emmambd Please advise how to properly name the notice and its description.

**Comment by smsm1-ito** [comment]
I'm thinking it's worth making this more generic by covering other similar text based fields that could also be affected, such as stop description, agency name, or route names.

**Comment by qcdyx** [comment]
@davidgamez We can create a new annotation, @NoReplacementChar, to cover the replacement char check for all related fields, and have a validator automatically generated. Should we still merge this PR, or delete the branch and open a new issue for the generic check? How would you like to proceed? @emmambd

**Comment by emmambd** [comment]
Thanks for chiming in @smsm1-ito! You make a great point. @qcdyx Let's delete this branch and create a generic notice instead.

**Comment by davidgamez** [comment]
Let's use the same issue.

**Comment by qcdyx** [comment]
Hey @emmambd and @davidgamez, any suggestions for the name of this annotation? @NoInvalidCharacters? @ShouldNotContainInvalidCharacters?

**Comment by qcdyx** [comment]
image

**Comment by qcdyx** [comment]
@emmambd @tzujenchanmbd May I have a list of fields' names and descriptions that require @NoInvalidCharacters decoration please?

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoInvalidCharactersNotice.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * </table>
+ */
+@GtfsValidationNotice(
+    severity = WARNING,
+    urls = {
+      @UrlRef(
+          label = "Best Practices for All Files",
+          url = "https://gtfs.org/schedule/reference/#file-requirements")
+    })
+public class NoInvalidCharactersNotice extends ValidationNotice {
```
Comment (davidgamez) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoInvalidCharactersNotice.java:33]: Rename the notice to:
```suggestion
public class InvalidCharactersNotice extends ValidationNotice {
```

Comment (qcdyx) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoInvalidCharactersNotice.java:33]: Good catch!

**Comment by qcdyx** [comment]
image

**Comment by skalexch** [comment]
@qcdyx @emmambd here are the text fields:
agencies: **agency_name**
stops: stop_code, **stop_name, stop_desc**, tts_stop_name, platform_code
routes: **route_short_name, route_long_name, route_desc**
trips: **trip_headsign, trip_short_name**
stop_times: **stop_headsign**
fare_media: fare_media_name
fare_products: fare_product_name
areas: area_name
networks: network_name
pathways: **signposted_as, reverse_signposted_as, level_name**
location_groups : **location_group_name**
booking_rules: **message, pickup_message, dropoff_message**
translations: field_name, **translation, field_value**
feed_info: field_publisher_name, field_version
attributions: organization_name

The ones in **bold** are the fields likely to have weird characters because the rest are usually numbers and basic latin letters (for codes and fare product names)and some are not likely public facing. Where do you think it's worth expanding to?

**Comment by jcpitre** [comment]
Question: @emmambd  Why do we need to annotate fields. Shouldn't this rule be applied to all fields?

**Comment by emmambd** [comment]
Based on outside discussion, it'll simplify the implementation to just apply this check generically to all strings. Let's do it universally and see the acceptance test results. cc @qcdyx @jcpitre

**Comment by qcdyx** [comment]
@davidgamez @emmambd @skalexch I've applied @NoInvalidCharacters to all String fields except ids, ready for a second review.

**Comment by davidgamez** [comment]
> @davidgamez @emmambd @skalexch I've applied @NoInvalidCharacters to all String fields except ids, ready for a second review.

We can address this more generically by adding the notice trigger while parsing the strings. This will make the notice future-proof and  remove the need for the annotation. See [link](https://github.com/MobilityData/gtfs-validator/blob/01b75393a03e4c0d83f0f45c8ce7e4682b3e36b2/core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java#L129).

**Comment by qcdyx** [comment]
@davidgamez The annotation approach makes the code cleaner and easier to maintain compared to manually checking row.asString(). Checking the entire row manually for invalid characters can consume more time compared to using the @NoInvalidCharacters annotation on individual string fields. Manual checking involves iterating through each character of the string and performing validation logic, which can be computationally expensive if done for every field in every row.

**Comment by davidgamez** [comment]
> @davidgamez The annotation approach makes the code cleaner and easier to maintain compared to manually checking row.asString(). Checking the entire row manually for invalid characters can consume more time compared to using the @NoInvalidCharacters annotation on individual string fields. Manual checking involves iterating through each character of the string and performing validation logic, which can be computationally expensive if done for every field in every row.

The _asString_ method doesn't check the entire row. It only parses the columns marked as String. If the intention is to check all string fields, this is the natural place. Otherwise, we will need to mark all string fields with the annotation, making the annotation redundant to the definition of the field type. I'm unsure what you meant by "manual check"; the exact implementation proposed by this PR(String.contains) can be placed in the asString method.

**Comment by qcdyx** [comment]
@davidgamez As discussed in standup, we close this PR and I'll work on a new branch for the asString() approach.

**Pull request closed without merging**